### DISCUSSION TITLE
use-package instructions

### DISCUSSION THREAD

**Discussion by al3xandru** [kickingvegas/casual-suite#28] [General]
Hi,

Firstly thanks for creating these projects.  I am a newbie Emacs user and feature discoverabiltiy is very useful for me.  I've also experimented with transient and I know it takes time to come up with something that makes sense.

I have looked at the installation instruction. As a newbie, I got some questions:

```
(require 'casual-suite)
(use-package casual-calc
  :ensure t
  :bind (:map calc-mode-map ("C-o" . casual-calc-tmenu)))
```

The `casual-suite` has to come from somewhere as it is not available locally.

So I think, it should start like:

```
(use-package casual-suite
  :ensure t)
```

Then as casual-calc is placed after the `(require 'casual-suite)`, 
I suppose the idea is to load it after casual-suite. 
Thus I'm thinking:

```
(use-package casual-calc
  :after casual-suite
  ...)
```

Similarly for the other ones requiring re-builder, ibuffer, etc.

What do you think?

**Comment by kickingvegas** [comment]
@al3xandru Thanks for sharing this. First off, let me confess that until this morning, I had a half-baked understanding of how `use-package` works, as I’ve largely avoided using it and instead explicitly declared `require` for packages. Then I finally watched [Protesilaos Stavrou’s](https://youtu.be/RaqtzemHFDU?si=if7FtrA5xg3tQOGF) video on `use-package` and with that information I can hopefully answer your questions.

Some observations:

-   The primary design intent of `use-package` is to load and configure packages.
-   It can also be used to install packages with `:ensure t`, delegating by default the installation to `package.el`.
-   The primary design intent of `casual-suite` is to be an umbrella installation package for all the Casual porcelains.
    -   Installing `casual-suite` will trigger the installation of all the Casual porcelains [defined in it](https://github.com/kickingvegas/casual-suite/blob/111e67a3bd548aa9ff2b2895602d0e7241a104f7/lisp/casual-suite.el#L9).
    -   The porcelains themselves have *no* dependent functionality on `casual-suite`.
-   Because `use-package` co-mingles configuration with installation, it can create confusion as evidenced by this conversation.

There’s a couple of scenarios I can see the user go through which I’ll enumerate below.


## User explicitly installs `casual-suite` via `package-list`

In this scenario, all Casual porcelains are already installed. There is no need to invoke `(require 'casual-suite)`. Each individual porcelain can be configured using `use-package`. The setting `:ensure t` is effectively a nop.


## User uses `use-package` to install `casual-suite`

In this scenario, if the user elects to use `use-package :ensure t` to install `casual-suite` then that is all it does. None of the Casual porcelains have any functional dependency on `casual-suite`, so `use-package` to configure any porcelain has no dependency on `casual-suite`.

This scenario seems the most problematic as ordering between installation (particularly of dependent packages) and loading is co-mingled. Also as `casual-suite` pulls in all the porcelains (effectively being a G-d library), it violates the design intent of `use-package` to facilitate lazy loading. I would avoid it.


## User sets `:ensure t` in `use-package` for any porcelain

In doing this, the need for invoking `use-package casual-suite` is obviated. Users who elect to do this don’t need to install `casual-suite`.

Note that in any of the scenarios above, `use-package` as of Emacs 29 will *not* auto-update a previously installed package. Apparently Emacs 30 will introduce a `:vc` keyword which will enable auto-updating.

Hopefully this provides some clarification. Upon reflection, `casual-suite` is really just a convenience package to install all the porcelains and should not be pushed to do more as suggested in #19 if it really wants to play nice with `use-package`.

**Comment by kickingvegas** [comment]
@al3xandru I've recently updated the README.org file with revised instructions for using `use-package`. Please give it a try and provide feedback. 

https://github.com/kickingvegas/casual-suite

**Comment by al3xandru** [comment]
@kickingvegas 

Thanks for clarifying the relationship between the individual casual(s) and casual-suite. 

At a quick glance the updated readme looks like the snippet I have used myself (with the exception of the extras I had in there for casual-suite). I'm removing those now and retrying.

**Reply by kickingvegas** [reply]
Let me know how this works out. Thanks!

**Reply by al3xandru** [reply]
I have ended up with with the following:

```elisp
(use-package casual-dired
  :bind (:map dired-mode-map ("C-o" . casual-dired-tmenu)))

(use-package casual-bookmarks
  :after (bookmark)
  :bind (:map bookmark-bmenu-mode-map
              ("C-o" . casual-bookmarks-tmenu)
              ("S" . casual-bookmarks-sortby-tmenu)
              ("J" . bookmark-jump)))

(use-package casual-isearch
  :bind (:map isearch-mode-map ("C-o" . casual-isearch-tmenu)))

(use-package casual-ibuffer
  :after (ibuffer)
  :bind (:map ibuffer-mode-map
             ...))
```

I am not including `:ensure t` as I have `(setq use-package-always-ensure t)`.

If I'd use casual-suite, I would something similar but:

1. only have casual-suite be `:ensure t`
2. have all the individual packages:
    * come `:after (casual-suite)`
    * marked as `:ensure nil`
